### ISSUE TITLE
Xamarin.Android/ARM Performance Problems Mapper.Initialize

### ISSUE THREAD

**Issue by sebastianfilke** [AutoMapper/AutoMapper#2870]
We have a big performance drop while using automapper with Xamarin Android while initialization and Mapper Profiles (around 250-300 mappings).

I read this post on stackoverflow and thats exactly what we recognize: https://stackoverflow.com/questions/51822796/q-how-to-improve-performance-with-initializing-maps-with-automapper-on-arm-devi

But i strip it down and i recognize that the problem even starts with 1 simple mapping. `CreateMap<IAddressModel, AddressVm>();`. I was removing everything from my initilization and the drop is nearly the same.

### Mapping configuration

```csharp
var watch = Stopwatch.StartNew();

            Mapper.Initialize(cfg => {
                cfg.AllowNullCollections = true;
                cfg.CreateMissingTypeMaps = false;
                cfg.EnableNullPropagationForQueryMapping = false;
                cfg.CreateMap<byte[], byte[]>().ConvertUsing(new ByteArrayConverter());
                cfg.AddProfile<BusinessLogicMapperProfile>();
                cfg.AddProfile<DataMapperProfile>();
                cfg.AddProfile<UiMapperProfile>();
            });

            watch.Stop();

            _logger.Info("Time elapsed for Mapper.Initialize: " + watch.Elapsed);
```

### Expected behavior

Same Performance as Windows or iOS or nearly the same

### Actual behavior
Initilization from above:
Android: Time elapsed for Mapper.Initialize: 00:00:13.1686282 -> 1147% Performance drop
Windows: Time elapsed for Mapper.Initialize: 00:00:01.1472908

Initilization with only 1 mapping:
Android: Time elapsed for Mapper.Initialize: 00:00:01.3858050 -> 926% Performance drop
Windows: Time elapsed for Mapper.Initialize: 00:00:00.1495924

### Steps to reproduce

Only add one mapping to automapper and you will see the problems in android.

**Comment by lbargaoanu** [comment]
Try the [MyGet](http://docs.automapper.org/en/stable/The-MyGet-build.html) build.

**Comment by lbargaoanu** [comment]
I, for one, don't write code on Android. So feel free to investigate what's going on. And once you know _exactly_ what's wrong, we can discuss about fixing it.

**Comment by sebastianfilke** [comment]
The Problem is still there? It's for Xamarin Forms. Any ideas how to fix it? What can i do to investigate on this? Can you give some advices. We cannot really use it because our initialization is too long.

**Comment by lbargaoanu** [comment]
You have to profile the part that's slow and find out where that time is spent. Or provide a repro without Xamarin. And then we'll see what can be done about it.

**Comment by sebastianfilke** [comment]
Yeah i profiled it already to a minimal example. If you only have one Mapping definition in a profile the performance already starts to be extremely slow. if you only add empty profiles the performance is the same. and it doesn't matter if the mapping definition is a big one or a very simple one. it was a very simple one in my case.

> Initilization with only 1 mapping:
Android: Time elapsed for Mapper.Initialize: 00:00:01.3858050 -> 926% Performance drop
Windows: Time elapsed for Mapper.Initialize: 00:00:00.1495924

**Comment by lbargaoanu** [comment]
That's not what profiling means. You might want to have a more senior colleague help you with that.

**Comment by sebastianfilke** [comment]
You seems to be a very friendly guy. i know what profiling is. i will do your job now. maybe we will further use automapper or use alternatives or make it simply on our own. then there will be no closed issues, because of we don't do it.

**Comment by jbogard** [comment]
Related? https://groups.google.com/forum/#!topic/automapper-users/rSrvhrAFic8

**Comment by jbogard** [comment]
I got a repro from that poster for 1200 maps, just haven't had any time to take a look at it. Also I have no idea how to profile on Android, not sure if just looking at dotTrace in full .NET would highlight the same areas?

**Comment by lbargaoanu** [comment]
I believe it should be possible to get an idea about what's going on by looking at that trace on your computer.

**Comment by jbogard** [comment]
Yeah I don't even have an Android device, and I'm not sure how much the emulator will help.

@sebastianfilke in your profile your numbers are just the total time, which isn't helpful really to figure out where the time is actually spent. Do you have a trace that shows *where* the time is being spent? From something like ANTS or dotTrace?

**Comment by BMarinov** [comment]
@sebastianfilke I seem to be having the same issue. Did you find a workaround?
Even a simple instantiation of MapperConfig seems to be taking around 2 seconds.

**Disclaimer: DEBUG Mode, debugger attached to an x64 emulator.**

Nr | Stopwatch | Operation | Description
------------- | ------------ | ------------- | -------------
(1) | 0:00:00.0306 | No Automapper cfg Instantiation | Handle forced invalid operation exception in automapper assembly.
(2) | 00:00:02.1009 | (1) + Instantiate empty mapper cfg | `var cfg = new MapperConfiguration(c =>{});`
(3) | 00:00:06.1593 | (2) + Register a profile | `var cfg = new MapperConfiguration(c => { c.AddProfile<DtoProfile>(); })`
(4) | 00:00:06.0216 | (3) + register mapper delegate in container | container.Register<IMapper>(made: Made.Of(() => new Mapper(cfg)) /*dryioc specific stuff here*/ );

Take these with a grain of salt. I ran the application a bunch of times and in every step I executed the code from the previous one plus the snippet in the description. The times were consistent. 

@jbogard  I know that Xamarin Forms isnt the target audience, but there seems to be something fishy going on. Can we do something to further diagnose the issue?
